sycl/plugins/level_zero/usm_allocator.cpp

```diff
@@ -514,9 +516,24 @@ class USMAllocContext::USMAllocImpl {
   USMAllocImpl(std::unique_ptr<SystemMemory> SystemMemHandle)
       : MemHandle{std::move(SystemMemHandle)} {
 
-    Buckets.reserve(BucketSizes.size());
+    // Buckets for Host and Shared allocations will use the full
+    // set of buckets, starting at the page size of 64 bytes.
+    auto NumBuckets = BucketSizes.size();
+
+    // Buckets for Device allocations will use starting size of 512.
+    // This is because memory compression on newer GPUs makes the
+    // minimum granularity 512 bytes instead of 64.
+    auto MT = MemHandle->getMemType();
+    if (MT == SystemMemory::Device)
+      // Skip buckets between 64 and 512.
+      NumBuckets -= 6;
+
+    Buckets.reserve(NumBuckets);
```

Comment:
I suggest reserving full size and use "shrink_to_fit" to release unused entries.
```suggestion
    Buckets.reserve(BucketSizes.size());
```

Reply:
The bucket initialization is done in a generic routine called for Host, Device and Shared. It uses a single array of sizes. Any shrinking of the array would have to be done so we use the largest set of elements first, and then successively reduce the set of elements. That introduces an ordering on initialization of Host/Device/Shared.
It might be better to have separate arrays for the three memory types.

Reply:
I didn't mean you'd shrink `BucketSizes`, but `Buckets` vector initialized in the loop below.

Reply:
Also you don't need to "reserve" space in vector and can just "push_back", and then you don't need to shrink

Reply:
Even more: I don't see why you need the constexpr `BucketSizes` and why the sizes cannot be computed dynamically in the loop initializing `Bucket`. This is not a performance sensitive initialization.

Reply:
Yes, it can be done dynamically without an array of sizes. The current way is historical.